src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -52,67 +76,165 @@ public void AddDependencyLocation(string fullPath)
+                }
+                else
                 {
-                    loadedAssembly = existingAssembly;
+                    assemblyName = null;
+                    realPath = "";
+                    hasData = false;
                 }
             }
 
-            // Otherwise, load the assembly.
-            if (loadedAssembly == null)
+            if (hasData)
             {
-                loadedAssembly = LoadFromPathUncheckedImpl(fullPath);
+                return (assemblyName, realPath);
             }
 
-            // Add the loaded assembly to the path cache.
-            lock (_guard)
+            try
+            {
+                realPath = PreparePathToLoad(originalAnalyzerPath);
+                assemblyName = AssemblyName.GetAssemblyName(realPath);
+
+                lock (_guard)
+                {
+                    _analyzerAssemblyInfoMap[originalAnalyzerPath] = (assemblyName, realPath);
+                }
+
+            }
+            catch
             {
-                _loadedAssembliesByPath[fullPath] = loadedAssembly;
+                // The above can fail with the assembly doesn't exist because it's corrupted, 
+                // doesn't exist on disk or is a native DLL. Those failures are handled when 
+                // the actual load is attempted. Just record the failure now.
             }
 
-            return loadedAssembly;
+            return (assemblyName, realPath);
         }
 
-        protected ImmutableHashSet<string>? GetPaths(string simpleName)
+        /// <summary>
+        /// Return the best path for loading an assembly with the specified <see cref="AssemblyName"/>. This
+        /// return is a real path to load, not an original path.
+        /// </summary>
+        protected string? GetBestPath(AssemblyName assemblyName)
         {
+            if (assemblyName.Name is null)
+            {
+                return null;
+            }
+
+            ImmutableHashSet<string>? paths;
             lock (_guard)
             {
-                _knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths);
-                return paths;
+                if (!_knownAssemblyPathsBySimpleName.TryGetValue(assemblyName.Name, out paths))
+                {
+                    return null;
+                }
+            }
+
+            // Sort the candidate paths by ordinal, to ensure determinism with the same inputs if you were to have
+            // multiple assemblies providing the same version.
+            string? bestPath = null;
+            AssemblyName? bestName = null;
+            foreach (var candidateOriginalPath in paths.OrderBy(StringComparer.Ordinal))
+            {
+                (AssemblyName? candidateName, string candidateRealPath) = GetAssemblyInfoForPath(candidateOriginalPath);
+                if (candidateName is null)
+                {
+                    continue;
+                }
+
+                bool isMatch;
+#if NETCOREAPP
```

Comment:
It's a sensible idea, I'm not sure I want to lump it into this PR though. It will cause fairly significant rename on top of all the churn already present here. May want to push that off to a follow up so I can get this bug fix in.

Reply:
I went ahead and spiked this out here 

https://github.com/jaredpar/roslyn/commit/c3fa753cc36766f1787506c42f8ed0b9d050482a

I agree we should do this but I'm leery to add it to the current PR. Feel like it's complicated enough. Think it's better to get this merged then follow up with that change quickly after. Open to suggestions.